### PR TITLE
fix(required): correctly validate when required on non-input element …

### LINKED ISSUE #16830: md-select not working properly with Angular 1.7.7
**I'm submitting a ...**

- [x ] bug report
- [ ] feature request
- [ ] other 

**Current behavior:**

Angular 1.7.7 results in a problem for the Angular Material md-select directive when the directive has the required attribute, has a parent with an ng-if, and the md-select's ng-model is given an initial value.   The directive incorrectly flags a required error and the displayed select option shows a trailing asterisk.

**Expected / new behavior:**

The md-select should not flag a required error and the displayed selected option should not show an asterisk.

**Minimal reproduction of the problem with instructions:**

In Controller:
```
$scope.selectModel = 'option1';
$scope.showSelect = true;
```

In View:
```
<md-input-container ng-if='showSelect'>
  <label> Select </label>
  <md-select ng-model='selectModel' required>
    <md-option value='option1'> Option 1 </md-option>
    <md-option value='option2'> Option 2 </md-option>
  </md-select>
</md-input-container>
```

**AngularJS version:** 1.7.7

AngularJS Material version: 1.1.12

**Browser:** Chrome 71.0.3578.98 | Firefox 64.0.2 | Edge 42.17134.1.0 

**Anything else:**

screen shot 02-08-19 at 02 27 pm

screen shot 02-08-19 at 02 27 pm 001

### PR DESCRIPTION
**Pull request by Narretz**
…is surrounded by ngIf

Closes #16830

I've also ran this change against the ngMaterial unit tests, the uiBootstrap unit tests, the ionic unit tests, and manually tested with uiSelect.

**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

**What is the current behavior? (You can also link to an open issue here)**

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change?**

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [ ] Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated
- [ ] Fix/Feature: Tests have been added; existing tests pass

**Other information**:

### CODE REVIEW

File: src/ng/directive/validators.js [hunk 1]
```diff
@@ -68,15 +68,21 @@ var requiredDirective = ['$parse', function($parse) {
     require: '?ngModel',
     link: function(scope, elm, attr, ctrl) {
       if (!ctrl) return;
-      var value = attr.required || $parse(attr.ngRequired)(scope);
+      // For boolean attributes like required, presence means true
+      var value = 'required' in attr || $parse(attr.ngRequired)(scope);
```
Comment (gkalpak) [inline src/ng/directive/validators.js]: Nit: `attr.hasOwnPropetty('required')` would avoid looking up the prototype chain and be more consistent with similar checks in other places.

File: src/ng/directive/validators.js [hunk 2]
```diff
@@ -68,15 +68,21 @@ var requiredDirective = ['$parse', function($parse) {
...
 
-      attr.required = true; // force truthy in case we are on non input element
+      if (!attr.ngRequired) {
+        // force truthy in case we are on non input element
+        // (input elements do this automatically for boolean attributes like required)
+        attr.required = true;
```
Comment (gkalpak) [inline src/ng/directive/validators.js]: Shouldn't this be consistent with `value` above?

Comment (Narretz) [reply src/ng/directive/validators.js]: What do you mean? If ngRequired isn't set, we must be in the required path. This condition is just for non-input elements.

Comment (gkalpak) [reply src/ng/directive/validators.js]: NVM, I forgot that at this point either `attr.required` or `attr.ngRequired` are defined 👍

Comment (Narretz) [reply src/ng/directive/validators.js]: I forgot this about two times myself while working on this 😄

**Comment by kambleaa007** [comment]
looks like, all checks are passed but not merged

**Review by petebacondarwin — APPROVED** [review]

**Pull request merged**